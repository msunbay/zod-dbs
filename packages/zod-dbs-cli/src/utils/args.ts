export const getArgumentValue = (name: string): string | undefined => {
  const argv = process.argv.slice(2);

  const key = name.startsWith('--') ? name : `--${name}`;
  for (let i = 0; i < argv.length; i++) {
    const arg = argv[i];
    if (arg === key) return argv[i + 1];
    if (arg.startsWith(`${key}=`)) return arg.slice(key.length + 1);
  }
  return undefined;
};

export const hasArgument = (name: string): boolean => {
  const argv = process.argv.slice(2);
  const key = name.startsWith('--') ? name : `--${name}`;
  return argv.includes(key) || argv.some((arg) => arg.startsWith(`${key}=`));
};

export const isSilentMode = (): boolean => {
  return hasArgument('--silent');
};
