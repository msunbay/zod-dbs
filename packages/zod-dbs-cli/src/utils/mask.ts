const maskConnectionString = (connectionString: string): string => {
  if (!connectionString) return '';

  let masked = connectionString;

  // 1) Try URL-style masking: scheme://username:password@host ...
  // Works for any scheme (postgres, mysql, mongodb+srv, mssql, snowflake, etc.)
  try {
    // Only attempt URL parsing if it looks like a URL (has a scheme:)
    if (/^[a-zA-Z][a-zA-Z0-9+.-]*:/.test(masked)) {
      // If it contains authority (//), URL will preserve userinfo
      if (masked.includes('//')) {
        const url = new URL(masked);
        if (url.password) {
          url.password = '****';
        }
        // Normalize URL (might reorder params), then continue with further masking
        masked = url.toString();
      }
    }
  } catch {
    // Ignore parse errors; fall back to regex approaches below
  }

  // 2) Generic userinfo masking without full URL parse: mask username:password@ (any scheme)
  // Replace only the password portion after the first ':' in the userinfo segment
  masked = masked.replace(
    /(\/\/)([^\/?#@]*)(@)/,
    (_m, p1: string, userinfo: string, p3: string) => {
      const idx = userinfo.indexOf(':');
      if (idx === -1) return `${p1}${userinfo}${p3}`;
      const userPart = userinfo.slice(0, idx + 1); // include colon
      return `${p1}${userPart}****${p3}`;
    }
  );

  // 3) Mask sensitive query parameters in URLs: ?password=...&token=...
  masked = masked.replace(
    /([?&])(password|pwd|token|sessiontoken|secret|clientsecret|client[_-]?secret|accesskeyid|secretaccesskey|access[_-]?key|secret[_-]?key|api[_-]?key|accountkey)=([^&#]*)/gi,
    (_m, sep: string, key: string) => `${sep}${key}=****`
  );

  // 4) Mask DSN-style key=value pairs (e.g., mssql/oracle). Apply only on DSN-like strings
  // to avoid clobbering URL query strings.
  if (
    masked.includes(';') ||
    (!/^[a-zA-Z][a-zA-Z0-9+.-]*:/.test(masked) && !masked.includes('?'))
  ) {
    masked = masked.replace(
      /(\b(password|pwd|token|sessiontoken|secret|clientsecret|client[_-]?secret|accesskeyid|secretaccesskey|access[_-]?key|secret[_-]?key|api[_-]?key|accountkey)\s*=\s*)('(?:[^']*)'|"(?:[^"]*)"|[^;&\s'"\)]+)/gi,
      (_m, prefix: string, _key: string, value: string) => {
        if (value?.startsWith("'") && value.endsWith("'"))
          return `${prefix}'****'`;
        if (value?.startsWith('"') && value.endsWith('"'))
          return `${prefix}"****"`;
        return `${prefix}****`;
      }
    );
  }

  return masked;
};

export const maskSensitiveValue = (
  name: string,
  value: string | undefined
): string => {
  if (!value) return '';

  const lowerName = name.toLowerCase();

  if (
    lowerName.includes('password') ||
    lowerName.includes('secret') ||
    lowerName.includes('token')
  ) {
    return '****';
  }

  if (name === 'connectionString') {
    return maskConnectionString(value);
  }

  return value;
};
