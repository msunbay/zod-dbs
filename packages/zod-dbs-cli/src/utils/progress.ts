import mustache from 'mustache';
import ora from 'ora';

const PROGRESS_STATUS: Record<string, string> = {
  connecting: 'Connecting to database...',
  fetchingSchema: 'Fetching schema information...',
  generating: 'Generating {{total}} Zod schemas...',
  done: 'Done',
};

const silentProgressHandler = {
  onProgress: () => {}, // No-op in silent mode
  done: () => {},
  fail: () => {},
};

export const createProgressHandler = (silent?: boolean) => {
  if (silent) return silentProgressHandler;

  const spinner = ora();

  return {
    onProgress: (status: string, args?: unknown) => {
      if (spinner.isSpinning) spinner.succeed();
      spinner.start(mustache.render(PROGRESS_STATUS[status] || status, args));
    },
    done: () => {
      spinner.succeed();
    },
    fail: () => {
      spinner.fail();
    },
  };
};
