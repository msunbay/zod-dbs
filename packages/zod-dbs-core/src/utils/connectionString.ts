/**
 * Configuration for database connection.
 */
interface ZodDbsConnectionConfig {
  protocol?: string; // e.g., 'postgresql', 'mysql', etc.
  /** Database port (default: 5432) */
  port?: number;
  /** Database host (default: localhost) */
  host?: string;
  /** Database name to connect to */
  database?: string;
  /** Username for authentication */
  user?: string;
  /** Password for authentication */
  password?: string;
  /**
   * Optional schema name (e.g., 'public' for PostgreSQL)
   */
  schemaName?: string;
}

export const createConnectionString = (
  options: ZodDbsConnectionConfig
): string => {
  const {
    host = 'localhost',
    port,
    database,
    user,
    password,
    protocol = 'db',
  } = options;

  let connectionString = `${protocol}://`;

  if (user) {
    connectionString += user;
    if (password) {
      connectionString += `:${password}`;
    }

    connectionString += `@${host}`;
  } else {
    connectionString += host;
  }

  if (port) {
    connectionString += `:${port}`;
  }

  if (database) {
    connectionString += `/${database}`;
  }

  return connectionString;
};

/**
 * Parses database connection strings, e.g., 'postgresql://user:pass@localhost:5432/dbname'
 */
export const parseConnectionString = (
  connectionString: string
): ZodDbsConnectionConfig => {
  const url = new URL(connectionString);
  const user = url.username;
  const password = url.password;
  const host = url.hostname;
  const port = url.port ? parseInt(url.port, 10) : undefined;
  const database = url.pathname.slice(1); // Remove leading '/'
  const protocol = url.protocol.replace(':', ''); // Remove trailing ':'

  return {
    host,
    port,
    database,
    user,
    password,
    protocol,
  };
};
