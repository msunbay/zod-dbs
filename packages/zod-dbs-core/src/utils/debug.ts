import debug from 'debug';

export const enableDebug = () => {
  process.env.DEBUG = 'zod-dbs:*';
  debug.enable('zod-dbs');
  logDebug('Debug mode enabled');
};

export const logDebug = (formatter: any, ...args: any[]) => {
  const logger = debug('zod-dbs');
  logger(formatter, ...args);
};
