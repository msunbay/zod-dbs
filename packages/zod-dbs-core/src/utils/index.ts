export * from './connectionString.js';
export * from './debug.js';
export * from './error.js';
export * from './sql.js';
