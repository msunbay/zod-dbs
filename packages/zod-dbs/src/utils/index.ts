export * from './casing.js';
export * from './fs.js';
export * from './mustache.js';
export * from './singularize.js';
