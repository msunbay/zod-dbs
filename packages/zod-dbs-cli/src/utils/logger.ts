import chalk from 'chalk';
import { ZodDbsProvider } from 'zod-dbs-core';

import { maskSensitiveValue } from './mask.js';

export const logSetting = (name: string, value: string | boolean | object) => {
  let displayValue = value.toString();

  if (name === 'provider' && typeof value === 'object' && value !== null) {
    const provider = value as ZodDbsProvider;
    const name = provider.name ?? 'custom provider';

    displayValue = provider.displayName
      ? `${name} (${provider.displayName})`
      : name;
  }

  console.info(
    `- ${chalk.white(name)}: ${chalk.blue(maskSensitiveValue(name, displayValue))}`
  );
};

export const logAppName = (message: string) => {
  console.info(chalk.magenta(`\n${message}\n`));
};

export const logError = (message: string) => {
  console.error(chalk.red(`${message}`));
};

export const logWarning = (message: string) => {
  console.warn(`⚠️ ${message}`);
};
