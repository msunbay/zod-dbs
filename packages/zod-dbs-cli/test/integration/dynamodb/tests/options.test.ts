import { execSync } from 'child_process';

import { getCliPath } from '../../utils.js';

const cliPath = getCliPath();

it('outputs provider options', async () => {
  const output = execSync(`node ${cliPath} --provider dynamodb --help`, {
    stdio: 'pipe',
  });

  expect(output.toString()).toMatchSnapshot();
});
