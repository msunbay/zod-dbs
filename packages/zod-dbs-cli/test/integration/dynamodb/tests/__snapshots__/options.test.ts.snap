// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`outputs provider options 1`] = `
"Usage: zod-dbs [options]

Generates Zod schemas from database schema.

Options:
  -V, --version                         output the version number
  -h, --help                            display help for command

DynamoDB options:
  --region <value>                      AWS region for DynamoDB (default:
                                        "us-east-1")
  --endpoint <value>                    Override endpoint (e.g., for local
                                        DynamoDB)
  --sample-size <value>                 Number of items to sample per table when
                                        inferring schema (default: 50)
  -p,--provider <name>                  Database provider to use

Output options:
  -o,--output-dir <path>                Output directory for generated schemas
                                        (default: "./zod-schemas")
  --module-resolution <type>            Module resolution type for generated
                                        files (choices: "commonjs", "esm",
                                        default: "commonjs")
  --clean-output                        Clean output directory before generating
                                        schemas
  --exclude <regex>                     Exclude tables matching this regex
  --include <regex>                     Include only tables matching this regex
  --silent                              Suppress all console output
  --debug                               Enable debug logging (default: false)
  --json-schema-import-location <path>  Path to import JSON schemas
  --zod-version <value>                 Zod version to use (choices: "3", "4",
                                        "4-mini", default: "3")
  --no-case-transform                   Disable case transformations /
                                        conversions for generated schemas
  --no-singularization                  Disable singularization of type and enum
                                        names
  --no-coerce-dates                     Disable using z.coerce.date() for date
                                        fields in read schemas
  --no-stringify-json                   Disable JSON.stringify() on json fields
                                        in write schemas
  --stringify-dates                     Convert dates to ISO strings in write
                                        schemas
  --default-empty-array                 Provide empty arrays as defaults for
                                        nullable array fields
  --default-unknown                     Whether to use "unknown" instead of
                                        "any" for unresolved types
  --object-name-casing <type>           Casing for generated object/type names
                                        (choices: "PascalCase", "camelCase",
                                        "snake_case", default: "PascalCase")
  --field-name-casing <type>            Casing for field/property names in
                                        schemas & records (choices: "camelCase",
                                        "snake_case", "PascalCase",
                                        "passthrough", default: "camelCase")
"
`;
